### DISCUSSION TITLE
v15.0.0-alpha.1

### DISCUSSION THREAD

**Discussion by spawnia** [webonyx/graphql-php#1231] [General]
### Changed

- PHP version required: 7.4+
- Propagate error message and stack trace for why leaf value serialization failed
- Do not throw client safe `Error` when failing to serialize an Enum type
- Use native PHP types for properties of `Type` and its subclasses
- Throw `SerializationError` over client safe `Error` when failing to serialize leaf types
- Move debug entries in errors under `extensions` key
- Use native PHP types wherever possible
- Always throw `RequestError` with useful message when clients provide an invalid JSON body
- Move class `BlockString` from namespace `GraphQL\Utils` to `GraphQL\Language`
- Return string-keyed arrays from `GraphQL::getStandardDirectives()`, `GraphQL::getStandardTypes()` and `GraphQL::getStandardValidationRules()`
- Move complexity related code from `FieldDefinition` to `QueryComplexity`
- Exclude unused standard types from the schema
- Require lazy type loader to return `Type` directly without an intermediary callable
- Allow lazy type loader to return `null`
- Rename `ServerConfig` option `persistentQueryLoader` to `persistedQueryLoader`
- Call previously unused methods `EnumType::parseValue()` and `EnumType::parseLiteral()`
- Strongly type `PromiseAdapter::createRejected()` to require `\Throwable`
- Move members specific to `NamedType` out of `Type`: `$name`, `$description`, `$config`, `isBuiltInType()`, `assertValid()`
- Always convert recursively when calling `Node::toArray()`
- Make `Directive::$config['args']` use the same definition style as `FieldDefinition::$config['args']`
- Rename `FieldArgument` to `Argument`
- Make errors when parsing scalar literals more precise
- Change expected `QueryPlan` options from `['group-implementor-fields']` to `['groupImplementorFields' => true]` in `ResolveInfo::lookAhead()`
- Always convert promises through `PromiseAdapter::convertThenable()` before calling `->then()` on them
- Use `JSON_THROW_ON_ERROR` in `json_encode()`
- Validate some internal invariants through `assert()`
- `PromiseAdapter::all()` accepts `iterable`
- Throw if `Introspection::fromSchema()` returns no data
- Reorganize abstract class `ASTValidationContext` to interface `ValidationContext`
- Reorganize AST interfaces related to schema and type extensions
- Align `Utils::suggestionList()` with the reference implementation (#1075)
- Order schema topologically and according to the user-defined order, affects introspection and printing
- `GraphQL\Utils\AST::typeFromAST()` now needs a type loader callable instead of the Schema
- Do not change HTTP status code in `StandardServer`
- Use `"` instead of `"""` for single line descriptions
- Make `Helper::emitResponse()` private, use `Helper::sendResponse()`
- Emit unescaped UTF-8 from `StandardServer`
- Sync input value coercion with `graphql-js` reference implementation

### Added

- Improve extendability of validator rules
- Add tests for errors that occur when undeclared fields are passed in input
- Warn about orphaned object types
- Expose structured enumeration of directive locations
- Add `AST::concatAST()` utility
- Allow lazy input object fields
- Add validation rule `UniqueEnumValueNames`
- Add SDL validation rule `UniqueOperationTypes` (#995)
- Add ability to remove custom validation rules after adding them via `DocumentValidator::removeRule()`
- Allow lazy enum values
- Make `Node` implement `JsonSerializable`
- Add SDL validation rule `UniqueTypeNames` (#998)
- Add support for SDL validation to `KnownTypeNames` rule (#999)
- Add SDL validation rule `UniqueArgumentDefinitionNames` (#1136)
- Add `parseValue` config option to InputObjectType to parse input value to custom value object
- Add option `sortTypes` to have `SchemaPrinter` order types alphabetically
- Allow constructing `EnumType` from PHP enum
- Add `TypeInfo::getParentTypeStack()` and `TypeInfo::getFieldDefStack()`
- Include path to faulty input in coercion errors

### Optimized

- Use recursive algorithm for printer and improve its performance
- Use `foreach` over slower functions `array_map()` and `Utils::map()`

### Fixed

- Avoid QueryPlan crash when multiple $fieldNodes are present
- Clarify error when attempting to coerce anything but `array` or `stdClass` to an input object
- Allow directives on variable definitions
- Handle `null` parent of list in `ValuesOfCorrectType::getVisitor`
- Allow sending both `query` and `queryId`, ignore `queryId` in that case
- Preserve extended methods from class-based types in `SchemaExtender::extend()`
- Fix printing of empty types (#940)
- Clone `NodeList` in `Node::cloneDeep()`
- Calling `Schema::getType()` on a schema built from SDL returns `null` for unknown types (#1068)
- Avoid crash on typeless inline fragment when using `QueryComplexity` rule
- Avoid calling `FormattedError::addDebugEntries()` twice when using default error formatting
- Avoid calling defined functions named like lazily loaded types
- Show actual error in debug entries

### Removed

- Remove deprecated `Type::getInternalTypes()`
- Remove deprecated `GraphQL::execute()`
- Remove deprecated `GraphQL::executeAndReturnResult()`
- Remove deprecated experimental CoroutineExecutor
- Remove deprecated `FormattedError::create()` and `FormattedError::createFromPHPError()`
- Remove deprecated `GraphQL::setPromiseAdapter()`
- Remove deprecated `AST::getOperation()`
- Remove deprecated constants from `BreakingChangesFinder`
- Remove deprecated `DocumentValidator::isValidLiteralValue()`
- Remove deprecated `Error::formatError()` and `Error::toSerializableArray()`
- Remove deprecated `GraphQL::getInternalDirectives()`
- Remove deprecated `Schema::isPossibleType()`
- Remove deprecated methods from `TypeInfo`
- Remove deprecated `Values::valueFromAST()` and `Values::isValidPHPValue()`
- Remove deprecated public property access to `InputObjectField::$type`
- Remove deprecated public property access to `FieldDefinition::$type`
- Remove alias `GraphQL\Validator\Rules\AbstractQuerySecurity`, use `GraphQL\Validator\Rules\QuerySecurityRule`
- Remove alias `GraphQL\Validator\Rules\AbstractValidationRule`, use `GraphQL\Validator\Rules\ValidationRule`
- Remove alias `GraphQL\Utils\FindBreakingChanges`, use `GraphQL\Utils\BreakingChangesFinder`
- Remove `OperationParams` method `getOriginalInput()` in favor of public property `$originalInput`
- Remove `OperationParams` method `isReadOnly()` in favor of public property `$readOnly`
- Remove `Utils::withErrorHandling()`
- Remove `TypeComparators::doTypesOverlap()`
- Remove `DocumentValidator::isError()`
- Remove `DocumentValidator::append()`
- Remove `Utils::getVariableType()` in favor of `Utils::printSafe()`
- Remove warning for passing `isDeprecated` in field definition config
- Remove `WrappingType::getWrappedType()` argument `$recurse` in favor of `WrappingType::getInnermostType()`
- Remove `Type::assertType()`
- Remove `ListOfType::$ofType`, `ListOfType::getOfType()` and `NonNull::getOfType()`
- Remove option `commentDescriptions` from `BuildSchema::buildAST()`, `BuildSchema::build()` and `Printer::doPrint()`
- Remove parameter `$options` from `ASTDefinitionBuilder`
- Remove `FieldDefinition::create()` in favor of `new FieldDefinition()`
- Remove `GraphQL\Exception\InvalidArgument`
- Remove `Utils::find()`, `Utils::every()` and `Utils::invariant()`
- Remove argument `bool $exitWhenDone` from `StandardServer::send500Error()` and `StandardServer::handleRequest()`
- Remove `Schema::getAstNode()` in favor of `Schema::$astNode`
- Remove ability to override standard types through `Schema` option `types`, use `Type::overrideStandardTypes()`
- Remove `GraphQL\Utils\TypeInfo::typeFromAST()`, use `GraphQL\Utils\AST::typeFromAST()`
- Remove `StandardServer::send500Error()`, handle non-GraphQL errors yourself
- Remove `StandardServer::getHelper()`, use `new Helper`

<hr /><em>This discussion was created from the release <a href='https://github.com/webonyx/graphql-php/releases/tag/v15.0.0-alpha.1'>v15.0.0-alpha.1</a>.</em>

**Comment by mfn** [comment]
Is https://github.com/webonyx/graphql-php/pull/830 mentioned? (I wasn't able to locate it in the changelog).

I noticed a lot of my tests "break" because the `extensions.category` isn't part of the result set and thus breaks the assertions.

**Reply by spawnia** [reply]
I just amended the changelog in `master`, good catch. An upgrade guide for this is already present.

Anything else that affected you in a major way and would benefit from an explicit upgrade guide?

**Reply by mfn** [reply]
Still working my way through the failed tests of https://github.com/rebing/graphql-laravel/pull/953 😅

Just a minor thing I couldn't spot in the changelog:
- `\GraphQL\Validator\DocumentValidator::getRule` was changed so that it does not use a secondary lookup by prefixing the namespace: `$name = sprintf('GraphQL\\Validator\\Rules\\%s', $name);`
- it's this commit https://github.com/webonyx/graphql-php/commit/d36e80962bd82b2f515617de50e0c87c8049bc3c#diff-fa0403de31548d680d863a100fdc906ec53696abb9d1e8b0cbc92020b7cce9a2L255
- it contains an update to `CHANGELOG.md` but this one didn't seem persist into the current version in master
  `DocumentValidator` relates changes are mentioned, but AFAICS not this one

**Reply by spawnia** [reply]
I restored the changelog entry.

**Comment by mfn** [comment]
I see no replacement / ability anymore to retrieve the values for [`getMaxQueryDepth`](https://github.com/webonyx/graphql-php/commit/144860da7e1b27b043adf0f31621be9a1139b639#diff-c97b042b892d828298749cb3ef6f6665c132994e962a30d7db513651fab52207L95-L99) and [`getMaxQueryComplexity`](https://github.com/webonyx/graphql-php/commit/61cecb0db1f43f1f98e6dc4f01c03628ccb31efe#diff-caa449492a616db9d164f28760efa8449671011513accac67a545ab858ec6756L253-L256) from `\GraphQL\Validator\Rules\QueryDepth` and `\GraphQL\Validator\Rules\QueryComplexity` respectively.

Is this intentional?

graphql-laravel allows configuring them and [I used the getters in the test](https://github.com/rebing/graphql-laravel/blob/6b8c3702941be95f0ff44b57e4596e8f0ccf5580/tests/Unit/ConfigTest.php#L74-L80) to ensure that the config was applied correctly.

**Reply by mfn** [reply]
I think I have it mostly; except what I mentioned so far, all is working. The above ⬆️ though what be cool to figure out if we can re-add the getters or what other way I can use to ensure the classes have been correctly configured.

I also think I need to wait for the v15 official release to finalize https://github.com/rebing/graphql-laravel/pull/953 to get the integration tests working (they use composer to literally install official Laravel from scratch), because the way I worked around in `composer.json` to have this v15 version _and_ the https://github.com/laragraph/utils/pull/4 PR won't work with them.

**Reply by spawnia** [reply]
I readded the setters.

**Reply by mfn** [reply]
Thank you, for graphql-laravel I'm almost ready to roll it out; everything seems to work and I also took it as an opportunity for other refactorings.

---

I still need to sift through a private project, I've so many failing GraphQL tests mostly related to the change in the error structure that it's hard to separate them from other stuff.

One thing I noticed there is that `\GraphQL\Executor\Promise\Adapter\SyncPromise::$queue` is not accessible anymore, as it now is a static var inside `\GraphQL\Executor\Promise\Adapter\SyncPromise::getQueue`.
In our 5k+ GraphQL integration tests, after the complexity of our resolvers increased which also involved sometimes multiple nested `Deferred` to resolve, we encountered that sometimes this queue was having "left overs" from a previous test, which would suddenly get deferred resolved in an unrelated test and cause issues. We used to reset this property to `null` in phpunits `setUp` and never looked back (the randomness made it extremely hard to get the to the bottom of it and it not something which is triggered in production, only in tests). But we think we can work this around in another way to manually clear the returned `SplQueue`, if it still happens.

**Reply by mfn** [reply]
I finally got around testing it, the TL;DR: all is good!

Some observations:
- that `\GraphQL\Executor\Promise\Adapter\SyncPromise::$queue` is not public anymore, was only a slight annoyance
  We now need to make sure that in `tearDown` we dequeue possible leftovers elements manually. Still, just setting it to `null` was easier, but this equally works.
- The nesting of additional fields in `errors` into `extensions` hit a lot of parts, but it wasn't hard to adapt the code
- the added types on properties added a bit of friction when using manual scalar types, but was easy fixable
- interestingly we were hit by this assertion https://github.com/webonyx/graphql-php/blob/5ad059b288f5e21e77cc9dcc45a525bb929e77bc/src/Executor/ReferenceExecutor.php#L843-L848
  There was always code doing this check, but only the new version showed up an issue with our code => great work; really appreciate using assertion which are showing the issues when testing things but not breaking prod!
- I noticed that quite some error messages don't reference field and arguments anymore; seems a pity from DX perspective; example
  ![image](https://user-images.githubusercontent.com/87493/222928162-2cf13d90-be28-40cb-8bd0-313ba81d7caa.png)

… and that's it :) Awesome work 🙏🏼
